### PR TITLE
Add ability to configure network protocol and load balancer class for external service configs

### LINKED ISSUE #331: Ability to set port network protocol (TCP/UDP) and LoadBalancerClass in externalServices
### Type of question

Best Practices

### Support question

Hello. Thank you for adding new feature [set container port network protocol](https://github.com/konpyutaika/nifikop/pull/320). But is there any way to add ability to change port network protocol (TCP/UDP) in `externalServices`? We are using AWS ALB/NLB. Due to the impossibility of assigning a protocol in the NiFi configuration file, we applied the changes manually in yaml of the service. This led to the fact that after applying changes to nifi yaml configuration, NiFiKop operator unsuccessfully tries to apply changes and it gives an error:

`{"level":"error","time":"2023-12-11T07:59:17.457Z","caller":"controller/controller.go:329","msg":"Reconciler error","controller":"nificluster","controllerGroup":"nifi.konpyutaika.com","controllerKind":"NifiCluster","nifiCluster":{"name":"nifi","namespace":"nifi"},"namespace":"nifi","name":"nifi","reconcileID":"b5e0bdca-6f32-4491-9df4-6b8886160e67","error":"failed to reconcile resource: updating resource failed: Service \"nlb\" is invalid: spec.loadBalancerClass: Invalid value: \"null\": may not change once set","errorVerbose":"updating resource failed: Service \"nlb\" is invalid: spec.loadBalancerClass: Invalid value: \"null\": may not change once set\nfailed to reconcile resource\ngithub.com/konpyutaika/nifikop/pkg/resources/nifi.(*Reconciler).Reconcile\n\t/workspace/pkg/resources/nifi/nifi.go:196\ngithub.com/konpyutaika/nifikop/controllers.(*NifiClusterReconciler).Reconcile\n\t/workspace/controllers/nificluster_controller.go:145\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1197","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235"}`

Seems, NiFiKop operator also can't downscale pods because of that error.

### NiFiKop version

v1.5.0-release

### Golang version

-

### Kubernetes version

Client Version: v1.28.2
Server Version: v1.28.2-eks

### NiFi version

1.23.2

### PR DESCRIPTION
**Pull request by mh013370**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | fixes #331 
| License         | Apache 2.0

### What's in this PR?

Adds the ability to set the network protocol for external services. Defaults to `TCP` and behaves almost identically to the recent addition to support this for internal listeners in #320.  

It also adds the ability to set the `loadBalancerClass` for external services. This is useful when using cloud based load balancers.

Example:

```yaml
  externalServices:
    - name: "clusterip"
      spec:
        type: ClusterIP
        portConfigs:
          - port: 8080
            internalListenerName: "http"
          - port: 7182
            internalListenerName: "my-custom-listener"
            protocol: TCP
      metadata:
        annotations:
          toto: tata
        labels:
          titi: tutu
```

And to set a `loadBalancerClass`:

```yaml
externalServices:
    - name: "nlb"
      spec:
        type: LoadBalancer
        loadBalancerClass: "service.k8s.aws/nlb"
        portConfigs:
          - port: 8080
            internalListenerName: "http"
          - port: 7182
            internalListenerName: "my-custom-listener"
            protocol: TCP
      metadata:
        annotations:
          toto: tata
        labels:
          titi: tutu
```

### Why?

It's possible to set the network protocol for container ports but not for services, so this just completes the feature.

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [x] Append changelog with changes

### CODE REVIEW

**Review by juldrixx — CHANGES_REQUESTED** [review]
Can you add in the example `loadBalancerClass` as you did for `protocol`?

**Comment by mh013370** [comment]
> Can you add in the example `loadBalancerClass` as you did for `protocol`?

In the docs? Surely!

File: site/docs/5_references/1_nifi_cluster/7_external_service_config.md [hunk 1]
```diff
@@ -14,6 +14,30 @@ ListenersConfig defines the Nifi listener types :
...
+            internalListenerName: "my-custom-listener"
+            protocol: TCP
+      metadata:
+        annotations:
+          toto: tata
+        labels:
+          titi: tutu
+```
+
+Load balancer example :
```
Comment (juldrixx) [inline site/docs/5_references/1_nifi_cluster/7_external_service_config.md:27]:
```suggestion
Load balancer example:
```

And can you do it too on `ListenersConfig defines the Nifi listener types :`?

Comment (juldrixx) [reply site/docs/5_references/1_nifi_cluster/7_external_service_config.md:27]: Just some typo 😄

**Review by juldrixx — CHANGES_REQUESTED** [review]

File: site/docs/5_references/1_nifi_cluster/7_external_service_config.md (hunk 1 above)
Comment (mh013370) [reply site/docs/5_references/1_nifi_cluster/7_external_service_config.md]: Done

**Review by juldrixx — APPROVED** [review]

**Pull request merged**